ravel instead of flatten

ozak opened the pull request:
Should decrease memory usage since it does not return a copy of the
cost matrix. This might be important for large files like the ones I am processing, which are over 4GB. So each copy requires additional memory of the same size. Additionally, ``flatten`` is slower in creation than ``ravel``. Should not impact anything else.

JDWarner commented:
@ozak thanks for looking into this. I'm on board with the change, with two comments before merging:

* an `.astype()` call occurs earlier in the same line and, as written, guarantees a copy. So while moving to `ravel` is a good call, if we want to actually avoid copying we also must set `.astype(..., copy=False)` in this line. Even then a copy may still be made.
* If we succeed in fully avoiding copies, the `self.flat_costs` attribute will be directly linked to the input array `costs`. In this case we must guarantee `self.flat_costs` is never modified internally, or those changes will silently affect the input.

Credit where credit is due - AppVeyor beat TravisCI done, which I've never seen before!

ozak commented:
Good catch!

I do not see why one would change the cost raster. The only thing that might change the values from what I see in the code is the heap, which I do not fully understand how it works, but I think that one just keeps indices and not the values of the costs. If that is the case then this should work.

ozak commented:
Hmmm this is failing on Travis. It passes in Travis on 2.6, 3.2, 3.4, but fails on 2.7 and 3.3. It complains that 

    TypeError: astype() takes no keyword arguments

due to some Numpy version issue perhaps? I had tried it on my computer and it worked ok. 

Regarding point 2 above I ran a test using a computer with the ``master branch`` and one with the ``ravel branch`` and they generated the same results. So I think my reading of the code seems correct and it should be safe to use ``ravel`` here.

blink1073 commented:
Keyword arguments to `astype()` are the problem.  You could wrap it in a try/except block to allow folks with the new numpy to benefit while still being backwards-compatible.

ozak commented:
Ok. Now it works perfectly.

stefanv commented on skimage/graph/_mcp.pyx:
```diff
@@ -299,7 +299,10 @@ cdef class MCP:
-        self.flat_costs = costs.astype(FLOAT_D).flatten('F')
+        try:
+            self.flat_costs = costs.astype(FLOAT_D, copy=False).ravel('F')
+        except:
```
Could you please add the Exception type here?  E.g. ``except ValueError``

ozak commented:
Not sure why this keeps failing in 2.7. The only thing I included is the TypeError, which is what I got when it had failed originally (see above). But now it seems it is only printing the message, but not returning the ``self.flat_costs`` raster? I had played with this on my computer and it worked fine. Any ideas?

stefanv commented on skimage/graph/_mcp.pyx:
```diff
@@ -299,7 +299,11 @@ cdef class MCP:
+            self.flat_costs = costs.astype(FLOAT_D, copy=False).ravel('F')
+        except TypeError:
+            self.flat_costs = costs.astype(FLOAT_D).flatten('F')
+            print('Using older Numpy version. Upgrading might decrease memory usage and increase speed.')
```
It may be that it is not a TypeError, in which case none of these statements will execute.

ozak commented:
The first time it failed the error message was a ``TypeError``, which is why I put that one. Any other way to figure out the type of the error?

stefanv commented on skimage/graph/_mcp.pyx:
```diff
@@ -299,7 +299,12 @@ cdef class MCP:
-        self.flat_costs = costs.astype(FLOAT_D).flatten('F')
+        try:
+            self.flat_costs = costs.astype(FLOAT_D, copy=False).ravel('F')
+        except TypeError, ex:
```
```
except Exception as ex
```

is what you want for testing purposes.  And then you can narrow it down a bit more.

ozak commented:
So, I changed the exception command to see what it said and I got the following in the tests:

    Expected:
        ([(0, 0), (0, 1), (1, 1)], 16.0)
    Got:
         <type 'exceptions.TypeError'>
        Error:  astype() takes no keyword arguments
        ('astype() takes no keyword arguments',)

Which means it is indeed a TypeError as I had written originally (or am I missing something?), so I do not know why it still failed in 2.7. Any ideas?

stefanv commented:
Indeed, that looks like it should work.

ozak commented:
Any other ideas as to why it fails?

jni commented:
It's because the 2.7 build uses NumPy 1.6, which does not allow keyword arguments:

http://docs.scipy.org/doc/numpy-1.6.0/reference/generated/numpy.ndarray.astype.html

ozak commented:
@jni that is true, but what is failing is the exception which is a ``TypeError`` and is being handled that way and still causing 2.7 to fail.

jni commented:
Oh. Sorry, this is what happens when you jump into an issue without reading all the background! =)

Incidentally, I don't know how Cython is keeping up to date with Python 3 syntax but that print statement made me cringe — can you change it to `print()`? Better yet, raise a warning.

jni commented:
OH! I just saw the error log. The print statement is messing with the doctests, which work by comparing output. Using the warnings module should fix this.

ozak commented:
Thought of doing that, using the ``warnings`` module, but wasn't sure if it would require installation, and changes in the ``requirements.txt`` file.

jni commented:
`warnings` is part of the standard library and is used in lots of other places in scikit-image. So, jump right in! =)

jni commented:
You *might* need to change some tests though, because we check for warnings during testing. Do a `git grep expected_warning` to see some usage examples.

ozak commented:
Thanks! Will do!

ozak commented:
Ok...now it works on 2.7 but fails everywhere else. Do I need an ``try`` and ``except`` statement in the tests also to make it work? Not sure what it should be in that case.

ozak commented:
How do can I put an `if`` statement using the Numpy version? I guess that would be the solution for the tests...

jni commented:
@ozak there is a way to make the expected warning optional:

https://github.com/scikit-image/scikit-image/blob/master/skimage/_shared/_warnings.py#L96

Sorry, should have pointed that out at the start! =)

jni commented:
btw, you don't need the full warning in there, just a key substring that is distinctive to your warning. =)

ozak commented:
Hmmm...not sure I understand how the optional part works...I'll search again all the tests and see. For now I hope the latest change should work on all (previous two worked on either 3.x and 2.6 or on 2.x so I joined both versions to cover all cases...not too clean for my taste).

jni commented:
@ozak, `with expected_warnings(['hello'])` means, there should be a warning raised containing the string 'hello'. `with expected_warnings(['hello|\A\Z'])` means, there should be that warning or none at all. That's because it's using regular expressions to match the warning, and that regex corresponds to 'hello or nothing', because `\A` is the start of a string and `\Z` is the end.

It's ugly syntax but there's little we can do about it! To [quote](https://twitter.com/iamdevloper/status/580326082269843456) I am devloper: "I’d really hate to see what Irregular Expressions look like…"

Anyway, looks like your current tests are failing because you need to cast the numpy version info, a character, as an integer type.

ozak commented:
So you mean I just needed to put the ``|\A\Z`` in the expected warning in [error tests](https://github.com/ozak/scikit-image/commit/454d63b0ece17d9114e3990a85eb5295465603f9)? That would make it much cleaner.

jni commented:
@ozak indeed. =)

I just thought of something that might make it a bit clearer:

```python
optional = r'|\A\Z'
with expected_warnings(['Upgrade NumPy' + optional]):
    m = ...
```

ozak commented:
Hmmm back to square one? The [Numpy Int version](https://github.com/ozak/scikit-image/commit/4f6e6d0bf6b65211f28b62c9b5ecaaa00960091a) worked on all (code not so nicely implemented). The ``Clean test`` version fails again in 2.7. Not sure if I understood what you meant @jni Juan or I just implemented it erroneously...did I still need to have the ``Numpy.version.version`` check or something?

jni commented:
@ozak yes, you did misunderstand. =) The string given to `expected_warnings` needs to match the actual warning string. My string was just an example because I didn't bother to look up the exact wording of your warning. In your case, it should be "Upgrading Numpy" rather than "Upgrade NumPy". (Apparently we write NumPy with a capital P in text; I've been reprimanded in the past for not doing that. =)

ozak commented:
@jni thanks for guiding me! Who would have thought such a simple tweak would generate all this back and forth :) Let's see if the next one works.

jni commented:
Haha I agree! I'm sitting here halfway between highly amused and apologetic! =D

ozak commented:
Success! Finally! Hahaha

stefanv commented:
@jni Just because I had nothing better to do ;)

jni commented:
@ozak Thank you very much for this. I think we want two things before merging:

- you only need one "optional" string, maybe right under the warning import, and then use that everywhere in the test file (rather than redefining it everywhere).
- we probably want to squash all these commits into one. Sorry, I know it's annoying to "lie" about how hard this was, but history will thank us! Let me know if you need help with squashing...!

ozak commented:
@jni I'll do as you command..hehehe

Not sure what you mean in the first point. I'll try to squash the commits after the change you propose...

thanks for all the help, and sorry to keep asking for more...it's been a nice learning experience!

jni commented:
In the test file, you define `optional = r'|\A\Z'` lots of times, in front of each `expected_warnings` call. Just define it once at the top of the module and use it everywhere. Perhaps, since it will be created "out of context", you can rename it to `warning_optional`.

ozak commented:
Ok...done as requested...I noticed it is failing, but it seems to be due to Numpy compilation error in AppVeyor. And in 3.3 due to some other test, which does not seem to be related to what was done here, since the error is

        ERROR: skimage.io.tests.test_pil.TestSaveTIF.test_imsave_roundtrip(<class 'bool'>, array([[[ True,  True,  True],
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/travis/venv/lib/python3.3/site-packages/nose/case.py", line 198, in runTest
        self.test(*self.arg)
      File "/home/travis/build/scikit-image/scikit-image/skimage/io/tests/test_pil.py", line 238, in roundtrip
        imsave(fname, x, compress=compress)
      File "/opt/python/3.3.5/lib/python3.3/contextlib.py", line 55, in __exit__
        next(self.gen)
      File "/home/travis/build/scikit-image/scikit-image/skimage/_shared/_warnings.py", line 112, in expected_warnings
        raise ValueError('Unexpected warning: %s' % str(warn.message))
    ValueError: Unexpected warning: unclosed file <_io.BufferedReader name='/home/travis/build/scikit-image/scikit-image/skimage/data/multipage_rgb.tif'>

Not sure if when I squashed and rebased it merged with some things that were not ready from somewhere else?

JDWarner commented:
@ozak The TravisCI error is definitely not your fault. It's an infrequent error that isn't confined to 3.3; I restarted that build and expect it will pass.

JDWarner commented:
@blink1073 what's the procedure for restarting an AppVeyor build? Must we close and re-open the PR for AppVeyor, like we once had to for Travis?

blink1073 commented:
@JDWarner, it appears to be a permissions issue.  For my own builds, I can trigger a new build, but not for the main repo ones.  Since @stefanv is away from keyboard, it seems a new commit is the only recourse.

stefanv commented:
How do I give permission to others to retrigger builds?

blink1073 commented:
@stefanv, when I pull up my own Projects, there is a settings icon, and then there is a Permissions tab.

stefanv commented:
Is that on Travis-CI?  I don't see such a tab.

blink1073 commented:
No, on Appveyor. 



—
Sent from Mailbox

On Tue, May 26, 2015 at 10:05 PM, Stefan van der Walt
<notifications@github.com> wrote:

> Is that on Travis-CI?  I don't see such a tab.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scikit-image/scikit-image/pull/1483#issuecomment-105740483

stefanv commented:
OK, try now.

ozak commented:
Am I supposed to be able to restart it now? Or is this something only you guys can do?

JDWarner commented:
@ozak I still can't restart AppVeyor, but we know what caused that error and a temporary fix has been applied.

Unfortunately, it's upstream of your PR. It will only get applied if you rebase so the fix is included. You will need to rebase your pull request on the current, upstream master. When you commit the rebase here, all the builds will restart but only AppVeyor is what we need checked again (assuming no further changes).

jni commented:
Do we really need to wait for a rebase to merge this? It's a pretty simple fix and passes everywhere else...

JDWarner commented:
Technically we're supposed to. I agree it's probably fine.

jni commented:
I'm happy to deal with the fallout should there be any. ;) @ozak thank you for the fix, and for sticking with it through all the failures! =)

ozak commented:
No problem. Happy to help out and also learn.

I might ask you some additional questions in a couple of days.
